proposals/fleet-multiple-abm-vpp.md

```diff
@@ -0,0 +1,145 @@
+### Multiple ABM/VPP tokens
+
+- Figma: https://www.figma.com/design/j2M1heOh8eZD6LcUJks6HE/%239956-Add-multiple-Apple-Business-Manager-and-Volume-Purchasing-Program-connections?node-id=2-130&t=3vPRYgndVgzoD8e4-0
+- Issue: https://github.com/fleetdm/fleet/issues/9956
+
+- Constraints/bussines logic:
+    - VPP tokens can be assigned to: "All teams", "No team" or a team. Default is "All teams"
+    - ABM tokens can be assigned to: "No team" or a team. Default is "No team"
+    - Multiple ABM/VPP tokens per team.
+    - Same private key can be used for multiple tokens.
+
+### ABM: gameplan overview
+
+The ADE workflow is summarized here
+
+https://github.com/fleetdm/fleet/blob/0a2a48b6d89ab9e428509eb918724c60c245db60/tools/mdm/apple/glossary-and-protocols.md?plain=1#L216-L235
+
+Currently, Fleet manages a single token, and it does the process described there in a cron job every 30 seconds.
+
+To support multiple tokens, we need to:
+
+- On each cron run, do the process for each token that was uploaded.
+- For each host ingested via ABM, keep track of the token that was used to ingest it.
+- Every time we need to make an ABM call to assign an ADE profile, choose the right token to use for the host.
+
+### VPP: gameplan overview
+
+The VPP token is used to:
+
+- Get a list of available apps for a team
+- Assign an app to a host
+
+To support multiple tokens, we need to:
+
+- Add a form to retrieve all tokens available for a team or "no team" this is `(team OR no team) ∪ all teams`.
+- To get a list of available apps: loop over each token, get and dedupe the available apps.
+- To assign an app to a host: loop over each token until we find the first one with available licenses, use it to assign the app to the host.
+- Track the token used to assign the app to the host.
+
+
+### Database migrations
+
+- Ticket: https://github.com/fleetdm/fleet/issues/21176
+
+#### Migrations
+
+**Tracking team associations for ABM/VPP tokens**
+
+Many-to-many relationship to be tracked in a join table:
+
+```sql
+CREATE TABLE mdm_asset_assignments (
```

Comment:
One thing that this fails to capture, if I read the Figma correctly, is that the same ABM token can be assigned to different teams based on the host's platform (!): https://www.figma.com/design/j2M1heOh8eZD6LcUJks6HE/%239956-Add-multiple-Apple-Business-Manager-and-Volume-Purchasing-Program-connections?node-id=5316-2520&t=anBMV3xXIbMrBv64-0

https://www.figma.com/design/j2M1heOh8eZD6LcUJks6HE/%239956-Add-multiple-Apple-Business-Manager-and-Volume-Purchasing-Program-connections?node-id=5316-2750&t=anBMV3xXIbMrBv64-0

Unless that's something we decided to cut for the first version?

Reply:
ohhh, great catch, thanks!! what do you think about a new column on this table? we'd still need to add three rows 🤔

Reply:
Yeah, maybe a `target_platform` ? And if `NULL`, is not platform-specific (i.e. all other assets)?

Reply:
sounds good, added!